### PR TITLE
MP2 Hessians

### PR DESCRIPTION
**Pull request by kcpearce**
## Description
Adds MP2 Hessians

## What are your new additions? Please provide a brief list.
* **New Features**
  - [x] Adds MP2 Hessians

## Any questions for the community?
- [x] Since Psi4 does not have MP2 Hessians yet, does anyone have recommendations as to which other code (CFour, Gaussian, others?) I should compare to? 

## Status
- [x] Derive Equations
- [x] Implement RHF Hessians and check
- [x] Implement MP2 Hessians
- [x] Verify with MP2 Hessians from another code
- [x] Add test
- [x] Ready for review-and-merge

### CODE REVIEW

**Comment by loriab** [comment]
Cfour would be a good choice. Not going to be accuracy-affecting approximations, easy to get a Cartesian Hessian out of. Known for higher-order derivatives. Let me know if you want a reference run.

**Comment by dgasmith** [comment]
Cool! Thank you for your continued work on this.

**Comment by kcpearce** [comment]
@loriab I'm currently having issues getting my version of CFOUR to work. If you get a chance, can you give me a reference run with the following following input/options:

```Water MP2/STO-3G Hessian Test
O
H 1 R 
H 1 R 2 A    

R=1.1
A=104

*CFOUR(CALC=MP2,
BASIS=STO-3G,
VIB=ANALYTIC,
SCF_CONV=12,
PRINT=10)```

Full disclosure: there could be problems with my input file. I only want the analytic hessian, but can't find the appropriate keyword to get that by itself, so I went with a vib freq calc instead. I also can't find out the highest print level allowed, but I'd like to get as much info in the output as possible.

**Comment by loriab** [comment]
* make sure mol is optimized for level of theory
* make sure there's a blank or comment line as first w/i cfour input
* can try some of these
  * `CALC_level=mp2`
  * `vibration=exact`
  * `FD_PROJECT=off`

**Comment by dgasmith** [comment]
@kcpearce This seems close, what else is needed before it is ready to go?

**Comment by kcpearce** [comment]
@dgasmith Just need to debug my t2 derivative code. Then it should be straightforward to get the hessians matching. And definitely a little cleaning up before review. Working on it now.

**Comment by dgasmith** [comment]
Awesome! I think this will make a really nice addition to the Psi4Numpy tutorials.

**Comment by dgasmith** [comment]
@kcpearce Looks like these are pretty close to ready to go?

**Comment by kcpearce** [comment]
@dgasmith Doing some further cleaning up of my code, but it's about ready to go! I'll get it pushed for review by the end of today.

**Comment by kcpearce** [comment]
@dgasmith Quick question: I'm using CFOUR's hessians to test mine, and I know we can use psi4 to interface to cfour for simple things (energies, gradients, etc.), but we can NOT do this for hessians, correct?

**Comment by loriab** [comment]
You can run hessians through the interface but the results aren’t parsed.

**Comment by kcpearce** [comment]
@loriab Okay, so I take it that I should just keep them hard coded in for my test at the end of the code?

**Comment by loriab** [comment]
Yes, hardcoding is the way to go here. Even in an analytic bs findif case, best to have a hard coded reference.

**Comment by kcpearce** [comment]
I ran pytest locally, and am failing `test_SAPT0_ROHF` on the final `psi4.compare_values` call. Any idea as to why?

**Comment by dgasmith** [comment]
Sorry no idea off the top of my head, how wrong is the answer?

**Comment by kcpearce** [comment]
`Ind200,r: computed value (-0.000251) does not match (0.000000) to 5 digits`

**Comment by dgasmith** [comment]
Ah pretty wrong... @loriab This looks like Psi4 changed the names of its variables?

**Comment by kcpearce** [comment]
Note: that test failure isn't caught by travis since it's marked as a long test and not run by travis

**Comment by dgasmith** [comment]
Can you make an issue about this then and we can go ahead and merge your PR when ready.

File: Moller-Plesset/MP2_Hessian.py [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+import copy
+import psi4
+from psi4 import *
+
+# Setup NumPy options
+np.set_printoptions(
+    precision=15, 
+    linewidth=200, 
+    suppress=True, 
+    threshold=np.nan
```
Comment (loriab) [inline Moller-Plesset/MP2_Hessian.py]: this is causing your error. replace `np.nan` with `np.maxsize`

Comment (kcpearce) [reply Moller-Plesset/MP2_Hessian.py]: Thanks. Took me a bit, but I just figured that out that's what was causing it almost simultaneously.

Comment (kcpearce) [reply Moller-Plesset/MP2_Hessian.py]: Hmm “AttributeError: module 'numpy' has no attribute 'maxsize'”

Comment (loriab) [reply Moller-Plesset/MP2_Hessian.py]: whoops, `sys.maxsize`

Comment (kcpearce) [reply Moller-Plesset/MP2_Hessian.py]: That fixed my 37 fail, but now I'm failing 36 for some reason. Is there a way to get Travis to show more useful output about why you're failing?

**Comment by dgasmith** [comment]
This is an elemental bug that I have just issues a patch for and not your fault. If this is ready to go we can review and merge.

**Comment by kcpearce** [comment]
Ready for review and merge on my end

File: Moller-Plesset/MP2_Gradient.py [hunk 2]
```diff
@@ -247,172 +235,92 @@
...
-Ppqrs[nocc:, nocc:, nocc:, :nocc] = Pabci
-
-# <VV|VV>
-# <--- sum_abcd ( [2 P(a,b,c,d) - P(a,b,d,c)] * <ab|cd> )
-Pabcd = 2 * Pabcd - Pabcd.swapaxes(2, 3)
-Ppqrs[nocc:, nocc:, nocc:, nocc:] = Pabcd
+# G'pqrs = Gpqrs + (2 * Ppr * kroecker_delta(q,occ) * kronecker_delta(q,s)) - (Pps * kronecker_delta(q,occ) * kronecker_delta(q,r))  
+for p in range(nmo):
+    for q in range(nmo):
+        for r in range(nmo):
```
Comment (dgasmith) [inline Moller-Plesset/MP2_Gradient.py]: Hmm, what about:
```
Ppqrs[:, :nocc, :, :nocc]  += 2.0 * np.einsum('pr,qs->pqrs', Ppq, np.eye(np.ones(nocc))))
```

or similar?

File: Moller-Plesset/MP2_Gradient.py [hunk 3]
```diff
@@ -429,96 +337,33 @@
...
-        #Gradient["J"][atom, p] += 4.0 * np.einsum('ciab,caib', Ppqrs[nocc:,:nocc,nocc:,nocc:], deriv1_np[map_key][nocc:,nocc:,:nocc,nocc:], optimize=True)
-        ## <OV|VV>
-        #Gradient["J"][atom, p] += 4.0 * np.einsum('icab,iacb', Ppqrs[:nocc,nocc:,nocc:,nocc:], deriv1_np[map_key][:nocc,nocc:,nocc:,nocc:], optimize=True)
-        ## <VV|OV>
-        #Gradient["J"][atom, p] += 4.0 * np.einsum('abic,aibc', Ppqrs[nocc:,nocc:,:nocc,nocc:], deriv1_np[map_key][nocc:,:nocc,nocc:,nocc:], optimize=True)
-        ## <VV|VO>
-        #Gradient["J"][atom, p] += 4.0 * np.einsum('abci,acbi', Ppqrs[nocc:,nocc:,nocc:,:nocc], deriv1_np[map_key][nocc:,nocc:,nocc:,:nocc], optimize=True)
-
-        ## <VV|VV>
-        #Gradient["J"][atom, p] += 4.0 * np.einsum('abcd,acbd', Ppqrs[nocc:,nocc:,nocc:,nocc:], deriv1_np[map_key][nocc:,nocc:,nocc:,nocc:], optimize=True)
```
Comment (dgasmith) [inline Moller-Plesset/MP2_Gradient.py]: I might have missed it, but why are these simplified so much?

File: Moller-Plesset/MP2_Hessian.py [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,prqs->', dPpqrs[key2], deriv1_np["TEI" + key1], optimize=True)
+        #
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tp,tqrs->', dPpqrs[key1], U[key2], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tq,ptrs->', dPpqrs[key1], U[key2], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tr,pqts->', dPpqrs[key1], U[key2], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,ts,pqrt->', dPpqrs[key1], U[key2], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tp,tqrs->', dPpqrs[key2], U[key1], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tq,ptrs->', dPpqrs[key2], U[key1], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,tr,pqts->', dPpqrs[key2], U[key1], npERI, optimize=True)
+        Hes["R"][r][c] += 0.5 * 1.0 * np.einsum('pqrs,ts,pqrt->', dPpqrs[key2], U[key1], npERI, optimize=True)
```
Comment (dgasmith) [inline Moller-Plesset/MP2_Hessian.py]: Wow, this is quite complex! I will go through this again in more detail soon.

File: .travis.yml [hunk 5]
```diff
@@ -44,8 +44,9 @@ before_script:
   - python -V
 
 script:
-  - python -m pytest -rs -v -m "not long" --cov --durations=5
+  #- python -m pytest -rs -v -m "not long" --cov --durations=5
   #- python -m pytest -rs -v -k "RHF"  # e.g., run only tests with RHF in name
+  - python -m pytest -rs -v -k "MP2_Hessian"
```
Comment (loriab) [inline .travis.yml]: switch this back at some point :-)

File: Moller-Plesset/MP2_Hessian.py [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+[-0.0000000000,       -4.2242489050,       -4.3809411906,        0.0000000000,        4.3722044693,        4.5463587882,       -0.0000000000,       -0.1479555642,       -0.1654175976],
+[-0.0000000000,       -4.7117763860,       -1.8779379264,        0.0000000000,        4.5463587882,        1.8072072616,        0.0000000000,        0.1654175977,        0.0707306648],
+[ 1.8065498535,        0.0000000000,        0.0000000000,        0.0691083574,       -0.0000000000,        0.0000000000,       -1.8756582110,       -0.0000000000,       -0.0000000000],
+[ 0.0000000000,       -4.2242489050,        4.3809411906,        0.0000000000,       -0.1479555642,        0.1654175976,       -0.0000000000,        4.3722044693,       -4.5463587882],
+[ 0.0000000000,        4.7117763860,       -1.8779379264,        0.0000000000,       -0.1654175977,        0.0707306648,       -0.0000000000,       -4.5463587882,        1.8072072616]
+])
+H_python_mat = psi4.core.Matrix.from_array(Hessian)
+psi4.compare_matrices(H_psi4, H_python_mat, 9, "RHF-HESSIAN-TEST")
```
Comment (loriab) [inline Moller-Plesset/MP2_Hessian.py]: `H_psi4` is from CFOUR? Good to add that `ZMAT` to the dir.

`RHF-HESSIAN-TEST` --> `RMP2-HESSIAN-TEST`?

File: Moller-Plesset/MP2_Hessian.py (hunk 4 above)
Comment (kcpearce) [reply Moller-Plesset/MP2_Hessian.py]: @dgasmith Let me know if you have any more comments/suggestions/changes with respect to the Hessian code

**Comment by kcpearce** [comment]
@dgasmith Good to go on my end!

**Pull request merged**